// @mastra/core/dist/signals/index.cjs
'use strict';

var chunkD37QL5LB_cjs = require('../chunk-D37QL5LB.cjs');

// src/signals/task-signal-provider.ts
var TaskSignalProvider = class extends chunkD37QL5LB_cjs.SignalProvider {
  id = "task-signals";
  #processor = new chunkD37QL5LB_cjs.TaskStateProcessor();
  getInputProcessors() {
    return [this.#processor];
  }
  getTools() {
    return {
      task_write: chunkD37QL5LB_cjs.taskWriteTool,
      task_update: chunkD37QL5LB_cjs.taskUpdateTool,
      task_complete: chunkD37QL5LB_cjs.taskCompleteTool,
      task_check: chunkD37QL5LB_cjs.taskCheckTool
    };
  }
};

Object.defineProperty(exports, "GoalSignalProvider", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.GoalSignalProvider; }
});
Object.defineProperty(exports, "SignalProvider", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.SignalProvider; }
});
Object.defineProperty(exports, "WebhookSignalProvider", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.WebhookSignalProvider; }
});
Object.defineProperty(exports, "isSignalProvider", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.isSignalProvider; }
});
exports.TaskSignalProvider = TaskSignalProvider;
