// @mastra/core/dist/llm/index.cjs
'use strict';

var chunkAWVVTLZF_cjs = require('../chunk-AWVVTLZF.cjs');
var chunkDURT53SS_cjs = require('../chunk-DURT53SS.cjs');
var chunkDBTKQ3U2_cjs = require('../chunk-DBTKQ3U2.cjs');
var chunkZMNN3DOV_cjs = require('../chunk-ZMNN3DOV.cjs');
var chunkOCS2GTEN_cjs = require('../chunk-OCS2GTEN.cjs');



Object.defineProperty(exports, "EMBEDDING_MODELS", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.EMBEDDING_MODELS; }
});
Object.defineProperty(exports, "ModelRouterEmbeddingModel", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.ModelRouterEmbeddingModel; }
});
Object.defineProperty(exports, "ModelRouterLanguageModel", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.ModelRouterLanguageModel; }
});
Object.defineProperty(exports, "defaultGateways", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.defaultGateways; }
});
Object.defineProperty(exports, "resolveModelAuth", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.resolveModelAuth; }
});
Object.defineProperty(exports, "resolveModelConfig", {
  enumerable: true,
  get: function () { return chunkAWVVTLZF_cjs.resolveModelConfig; }
});
Object.defineProperty(exports, "AzureOpenAIGateway", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.AzureOpenAIGateway; }
});
Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "MastraGateway", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.MastraGateway; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.parseModelString; }
});
Object.defineProperty(exports, "ModelsDevGateway", {
  enumerable: true,
  get: function () { return chunkDBTKQ3U2_cjs.ModelsDevGateway; }
});
Object.defineProperty(exports, "NetlifyGateway", {
  enumerable: true,
  get: function () { return chunkZMNN3DOV_cjs.NetlifyGateway; }
});
Object.defineProperty(exports, "GATEWAY_AUTH_HEADER", {
  enumerable: true,
  get: function () { return chunkOCS2GTEN_cjs.GATEWAY_AUTH_HEADER; }
});
Object.defineProperty(exports, "MastraModelGateway", {
  enumerable: true,
  get: function () { return chunkOCS2GTEN_cjs.MastraModelGateway; }
});
