// @mastra/core/dist/tools/index.cjs
'use strict';

var chunkD37QL5LB_cjs = require('../chunk-D37QL5LB.cjs');
var chunkER5YO3AZ_cjs = require('../chunk-ER5YO3AZ.cjs');
var chunkRS7FSLKM_cjs = require('../chunk-RS7FSLKM.cjs');
var chunk2TATDSHU_cjs = require('../chunk-2TATDSHU.cjs');



Object.defineProperty(exports, "DEFAULT_GOAL_JUDGE_PROMPT", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.DEFAULT_GOAL_JUDGE_PROMPT; }
});
Object.defineProperty(exports, "DEFAULT_GOAL_MAX_RUNS", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.DEFAULT_GOAL_MAX_RUNS; }
});
Object.defineProperty(exports, "FRAME_PREFIX", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.FRAME_PREFIX; }
});
Object.defineProperty(exports, "GOAL_STATE_ID", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.GOAL_STATE_ID; }
});
Object.defineProperty(exports, "GOAL_STATE_TYPE", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.GOAL_STATE_TYPE; }
});
Object.defineProperty(exports, "GoalStateProcessor", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.GoalStateProcessor; }
});
Object.defineProperty(exports, "StdioCodeModeTransport", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.StdioCodeModeTransport; }
});
Object.defineProperty(exports, "TASKS_REQUEST_CONTEXT_KEY", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.TASKS_REQUEST_CONTEXT_KEY; }
});
Object.defineProperty(exports, "TASKS_STATE_ID", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.TASKS_STATE_ID; }
});
Object.defineProperty(exports, "TASK_STATE_TYPE", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.TASK_STATE_TYPE; }
});
Object.defineProperty(exports, "TaskStateProcessor", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.TaskStateProcessor; }
});
Object.defineProperty(exports, "askUserTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.askUserTool; }
});
Object.defineProperty(exports, "assignTaskIds", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.assignTaskIds; }
});
Object.defineProperty(exports, "buildProgramModule", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.buildProgramModule; }
});
Object.defineProperty(exports, "buildRunner", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.buildRunner; }
});
Object.defineProperty(exports, "clearObjective", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.clearObjective; }
});
Object.defineProperty(exports, "createCodeMode", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.createCodeMode; }
});
Object.defineProperty(exports, "createCodeModeInstructions", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.createCodeModeInstructions; }
});
Object.defineProperty(exports, "createCodeModeTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.createCodeModeTool; }
});
Object.defineProperty(exports, "demoteExtraInProgress", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.demoteExtraInProgress; }
});
Object.defineProperty(exports, "formatQuestionAnswer", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.formatQuestionAnswer; }
});
Object.defineProperty(exports, "formatTaskListResult", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.formatTaskListResult; }
});
Object.defineProperty(exports, "generateStubs", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.generateStubs; }
});
Object.defineProperty(exports, "getObjectiveFromRequestContext", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.getObjectiveFromRequestContext; }
});
Object.defineProperty(exports, "getTasksFromRequestContext", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.getTasksFromRequestContext; }
});
Object.defineProperty(exports, "hasMultipleInProgress", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.hasMultipleInProgress; }
});
Object.defineProperty(exports, "jsonSchemaToTsString", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.jsonSchemaToTsString; }
});
Object.defineProperty(exports, "readObjective", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.readObjective; }
});
Object.defineProperty(exports, "resolveEffectiveGoalSettings", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.resolveEffectiveGoalSettings; }
});
Object.defineProperty(exports, "resolveGoalStore", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.resolveGoalStore; }
});
Object.defineProperty(exports, "submitPlanTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.submitPlanTool; }
});
Object.defineProperty(exports, "summarizeTaskCheck", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.summarizeTaskCheck; }
});
Object.defineProperty(exports, "taskCheckTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.taskCheckTool; }
});
Object.defineProperty(exports, "taskCompleteTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.taskCompleteTool; }
});
Object.defineProperty(exports, "taskUpdateTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.taskUpdateTool; }
});
Object.defineProperty(exports, "taskWriteTool", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.taskWriteTool; }
});
Object.defineProperty(exports, "writeObjective", {
  enumerable: true,
  get: function () { return chunkD37QL5LB_cjs.writeObjective; }
});
Object.defineProperty(exports, "ToolStream", {
  enumerable: true,
  get: function () { return chunkER5YO3AZ_cjs.ToolStream; }
});
Object.defineProperty(exports, "noopObserve", {
  enumerable: true,
  get: function () { return chunkER5YO3AZ_cjs.noopObserve; }
});
Object.defineProperty(exports, "MASTRA_TOOL_MARKER", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.MASTRA_TOOL_MARKER; }
});
Object.defineProperty(exports, "Tool", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.Tool; }
});
Object.defineProperty(exports, "createTool", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.createTool; }
});
Object.defineProperty(exports, "isProviderDefinedTool", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.isProviderDefinedTool; }
});
Object.defineProperty(exports, "isProviderTool", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.isProviderTool; }
});
Object.defineProperty(exports, "isValidationError", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.isValidationError; }
});
Object.defineProperty(exports, "isVercelTool", {
  enumerable: true,
  get: function () { return chunkRS7FSLKM_cjs.isVercelTool; }
});
Object.defineProperty(exports, "getTransformedToolPayload", {
  enumerable: true,
  get: function () { return chunk2TATDSHU_cjs.getTransformedToolPayload; }
});
Object.defineProperty(exports, "hasTransformedToolPayload", {
  enumerable: true,
  get: function () { return chunk2TATDSHU_cjs.hasTransformedToolPayload; }
});
