// @mastra/core/dist/fga-check-JNZ754RN.cjs
'use strict';

var chunkTKFSM674_cjs = require('./chunk-TKFSM674.cjs');



Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.checkFGA; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkTKFSM674_cjs.requireFGA; }
});
