// @mastra/core/dist/provider-registry-2MJBSB5W.cjs
'use strict';

var chunkDURT53SS_cjs = require('./chunk-DURT53SS.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isOfflineMode", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.isOfflineMode; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.isValidModelId; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkDURT53SS_cjs.parseModelString; }
});
